src/engine/sidechain/engineshoutcast.cpp

```diff
@@ -706,3 +704,99 @@ void EngineShoutcast::infoDialog(QString text, QString detailedInfo) {
+    m_readSema.release();
+}
+
+// Is called from the Mixxx engine thread
+void EngineShoutcast::setOutputFifo(FIFO<CSAMPLE>* pOutputFifo) {
+    m_pOutputFifo = pOutputFifo;
+}
+
+void EngineShoutcast::run() {
+    unsigned static id = 0;
+    QThread::currentThread()->setObjectName(QString("EngineShoutcast %1").arg(++id));
+    qDebug() << "EngineShoutcast::run: starting thread";
+
+    ignoreSigpipe();
+
+    DEBUG_ASSERT_AND_HANDLE(m_pOutputFifo) {
+        return;
+    }
+
+    setState(NETWORKSTREAMWORKER_STATE_BUSY);
+    if (!processConnect()) {
+        errorDialog(tr("Can't connect to streaming server"),
+                    tr("Please check your connection to the Internet and verify that your username and password are correct."));
+        return;
+    }
+
+    for(;;) {
+        m_readSema.acquire();
+        // Check to see if Shoutcast is enabled, and pass the samples off to be
+        // broadcast if necessary.
+        if (!m_pShoutcastEnabled->toBool()) {
+            m_threadWaiting = false;
+            m_pStatusCO->setAndConfirm(STATUSCO_UNCONNECTED);
+            processDisconnect();
+            return;
+        }
+        int readAvailable = m_pOutputFifo->readAvailable();
+        if (readAvailable) {
+            CSAMPLE* dataPtr1;
+            ring_buffer_size_t size1;
+            CSAMPLE* dataPtr2;
+            ring_buffer_size_t size2;
+            // We use size1 and size2, so we can ignore the return value
+            (void)m_pOutputFifo->aquireReadRegions(readAvailable, &dataPtr1, &size1,
+                    &dataPtr2, &size2);
+            process(dataPtr1, size1);
+            if (size2 > 0) {
+                process(dataPtr2, size2);
+            }
+            m_pOutputFifo->releaseReadRegions(readAvailable);
+        }
+    }
+}
+
+bool EngineShoutcast::threadWaiting() {
+    return m_threadWaiting;
+}
+
+void EngineShoutcast::ignoreSigpipe()
+{
+    // shout_send_raw() can cause SIGPIPE, which is passed to this theread
+    // and which will finally crash Mixxx if it remains unhandled.
+    // Each thread has its own signal mask, so it is safe to do this for the
+    // shoutcast thread only
+    // http://www.microhowto.info/howto/ignore_sigpipe_without_affecting_other_threads_in_a_process.html
+    sigset_t sigpipe_mask;
+    sigemptyset(&sigpipe_mask);
+    sigaddset(&sigpipe_mask, SIGPIPE);
+    sigset_t saved_mask;
+    if (pthread_sigmask(SIG_BLOCK, &sigpipe_mask, &saved_mask) == -1) {
+        qDebug() << "EngineShoutcast::ignoreSigpipe() failed";
+    }
+}
+
+
+void EngineShoutcast::slotStatusCO(double v) {
+    // Ignore external sets "status"
+    Q_UNUSED(v);
+    qWarning() << "WARNING:"
+            << SHOUTCAST_PREF_KEY << "\"status\" is a read-only control.";
+}
+
+void EngineShoutcast::slotEnableCO(double v) {
+    if (v > 1.0) {
+        // Wrap around manually .
+        // Wrapping around in WPushbutton does not work
+        // since the status button has 4 states, but this CO is bool
```

Comment:
It is on shade skin. 
It has two connections "enable" and "state"

> why not convert this to a multistate pushbutton?

The widget is already a multi state pushbutton. But since "enable" has a different count of states compared to the "state" (used as indicator connection), the button logic is confused. 
Finally this should be fixed in the widget. But since a good solution is not obvious and may introduce regressions this workaround works good for me.